lib/Doctrine/DBAL/Logging/LoggerChain.php

```diff
@@ -7,12 +7,22 @@
  */
 class LoggerChain implements SQLLogger
 {
-    /** @var SQLLogger[] */
+    /** @var iterable<SQLLogger> */
     private $loggers = [];
 
+    /**
+     * @param SQLLogger[] $loggers
+     */
+    public function __construct(iterable $loggers = [])
```

Comment:
It cannot be any iterable because it has to support `$this->loggers[] = $logger`. Please use `array` in `master`.

Reply:
can't do, `!tagged` does not use `array`. This is a requirement for lazyness

Reply:
Maybe throw an exception in the adder if `$this->loggers` is not an array then?

Reply:
sure if it seems worth it to you, @morozov would you like that?

Reply:
alternatively, `iterator_to_array` can be triggered from within `addLogger`

Reply:
What is `!tagged`?

Reply:
Symfony DI feature https://symfony.com/blog/new-in-symfony-3-4-simpler-injection-of-tagged-services

Reply:
Is this limitation of Symfony DI by design or just a shortcoming? It should be possible to inject an array of proxies which would instantiate the actual objects at the first call.

Why should loggers be lazily injected? They should be fairly lightweight by definition.

I don't like the hackery with `iterator_to_array` either. Does it mean adding a new logger will instantiate all existing one as a side effect?

Can we make it `array` in `2.x` and loosen to `iterable` in `3.0`?

Reply:
> It should be possible to inject an array of proxies which would instantiate the actual objects at the first call.

Proxies require classes to be non-final

> Why should loggers be lazily injected? They should be fairly lightweight by definition.

Feature doesn't distinguish it's dbal logger, it's universal for all services, so performance boost applies everywhere. And that's good, it doesn't hurt to apply it for lightweight services too.

> Does it mean adding a new logger will instantiate all existing ones as a side effect?

Yes. But remember, it's same case currently. Benefit of this approach would be that they are instantiated only when you use both methods - constructor and adder, which shouldn't really be done.

> Can we make it array in 2.x and loosen to iterable in 3.0?

What about deprecation path? With regards upgrade path, that doesn't seem correct to me. And I wanted to utilize this in doctrinebundle before dbal 2.0.

Reply:
> Proxies require classes to be non-final

DBAL doesn't depend on a specific final class. It depends on an interface which definitely can be proxied.

> Feature doesn't distinguish it's dbal logger, it's universal for all services, so performance boost applies everywhere. And that's good, it doesn't hurt to apply it for lightweight services too.

Proxying has its cost as well which may be even higher than no proxying for lightweight services which loggers usually are. So if we're talking about performance, let's look at the numbers.

> Yes. But remember, it's same case currently. Benefit of this approach would be that they are instantiated only when you use both methods - constructor and adder, which shouldn't really be done.

IMO it's too much complication for the reason I don't currently see. Instead of throwing an exception, one could use `AppendIterator` but I still don't think it's worth it.

> I wanted to utilize this in doctrinebundle before dbal 3.0.

If it's that important, you can add a temporary implementation with the logic you think fits and remove it once 3.0 is out.

Reply:
As you wish. I did the change you asked me to do. When is dbal 3.0 release scheduled?